src/compiler/transformers/declarations/coalesceImports.ts

```diff
@@ -0,0 +1,231 @@
+                    if (match) {
+                        return updateEntityNameRoot(node.qualifier!, !!match.name.autoGenerateFlags ? match.name : factory.createIdentifier(idText(match.name)));
+                    }
+                }) || factory.cloneNode(node.qualifier!);
+            }
+
+            function createOrUpdateImportClause(clause: ImportClause | undefined, name: Identifier | undefined, namedBindings: NamedImportBindings | undefined) {
+                if (!clause) {
+                    return factory.createImportClause(/*isTypeOnly*/ true, name, namedBindings);
+                }
+                return factory.updateImportClause(clause, clause.isTypeOnly, name, namedBindings);
+            }
+
+            function createOrUpdateNamedBindings(bindings: NamedImportBindings | undefined, specs: Identifier[], useLocations: Node[]) {
+                const specifiers = bindings && isNamedImports(bindings) ? [...bindings.elements] : [];
+                const specMap = new Map<string, ImportSpecifier>();
+                for (const spec of specifiers) {
+                    specMap.set(idText(spec.propertyName || spec.name), spec);
+                }
+                forEach(specs, s => {
+                    const renamed = some(useLocations, n => !!resolver.resolveName(idText(s), n.nextContainer || n, SymbolFlags.All, /*skipGlobals*/ false)) ? factory.createUniqueName(idText(s)) : undefined;
+                    if (!renamed ? specMap.has(idText(s)) : !!specMap.get(idText(s))?.name.autoGenerateFlags) return; // skip dupes (renamed ones can share a generated name)
+                    const spec = factory.createImportSpecifier(/*isTypeOnly*/ true, renamed ? s : undefined, renamed || s);
+                    specifiers.push(spec);
+                    specMap.set(idText(s), spec);
+                });
+                if (!bindings) {
+                    return factory.createNamedImports(specifiers);
+                }
+                return factory.updateNamedImports(bindings as NamedImports, specifiers);
+            }
+
+            function combineImportTypes(node: Node): VisitResult<Node> {
+                if (isImportDeclaration(node)) {
+                    // TODO: Since imports are hoisted (albeit with TDZ), we technically need to modify every import declaration _before_ recuring
+                    // on import type nodes to modify those, since technically the import an import type node relies on can occur _after_ it in
+                    // a file. This is somewhat unfortunate since it makes this require 3 tree walks - one to gather the imports and import type nodes we care about,
+                    // one (albeit likely shallow one) to modify the existing import declarations, and one to update all the import type nodes themselves.

```

Comment:
```suggestion
                    // TODO: Since imports are hoisted, albeit with a Temporal Dead Zone (TDZ), we technically need to modify every import declaration _before_ recurring on import type nodes.
                    // This is because the import an import type node relies on can occur _after_ it in a file.
                    // Unfortunately, this makes the process require three tree walks: one to gather the imports and import type nodes we care about, one (likely shallow) to modify the existing import declarations, and one to update all the import type nodes themselves.
```